Userland/Libraries/LibCore/Socket.h

```diff
@@ -193,9 +193,13 @@ class TCPSocket final : public Socket {
 
     virtual ~TCPSocket() override { close(); }
 
+    int fd() const { return m_helper.fd(); }
```

Comment:
Given the rationale: 

> Non-blocking use cases at the moment require direct usage of syscalls and knowledge of the fd. Mixing this with a TCPSocket is only convenient if we have TCPSocket::fd().

This feels like the wrong abstraction.

Instead of exposing internals of this class for meddling with socket and file descriptor options, we should provide helper methods like `set_nonblocking()` or `set_blocking(bool)` to expose a descriptive API describing the operations we want users to be able to do with the socket object

Reply:
This is a valid point, `TCPSocket::fd()` would indeed leak an implementation detail from a lower level.

I was considering withdrawing this PR and work on an async version of `TCPSocket`, or augment the existing `TCPSocket` by making it expose an interface suitable for any non-blocking socket usages.

Before doing so, I did some research to compare the approches taken by various libraries and observed that most if not all of them do explicitly leakily expose a method allowing the retrieval of the underlying socket fd:
* C++
  * Asio (Boost or standalone) 
    * https://stackoverflow.com/questions/58657896/how-to-get-native-socket-file-describer-from-boostasiosslstreamboostasi
    * https://www.boost.org/doc/libs/1_82_0/doc/html/boost_asio/reference/ip__tcp/socket.html  (lowest_layer, native_handle)
  * Qt
    * https://doc.qt.io/qt-6/qabstractsocket.html#socketDescriptor
* Rust
  * net
    * https://doc.rust-lang.org/std/net/struct.TcpStream.html#impl-AsFd-for-TcpStream
  * async_net
    * https://docs.rs/async-net/latest/async_net/struct.TcpStream.html#method.as_raw_fd
* Python
  * socket
    * https://docs.python.org/3/library/socket.html#socket.socket.fileno
  * asyncio
    * https://docs.python.org/3/library/asyncio-protocol.html#asyncio.BaseTransport.get_extra_info


I think one of the reason for those libraries to leak the fd is that sockets (blocking or not) are very finicky to work with and debug, even more so when the code using them is potentially ran atop of various networking stack implementations.
Java seems to be an exception here, probably because they respect their "write once, run anywhere" motto; so, no fd leak: https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/nio/channels/SocketChannel.html Which is funny because their "abstraction" is not much of an abstraction, it's the typical POSIX `select()` flow all over but in Java, we just don't have access to the fd.

Boost/Asio have the whole async/non-blocking socket use case well abstracted for async connects/reads/writes with callbacks using an event loop, yet the fd is still very much accessible.

So while we'd ideally have something like Asio eventually, I think there's no shame in making the fd accessible and that it might even be desirable.

As for my use case and the reason we ended up here: I started developing a BitTorrent client from scratch using SerenityOS/AK and quickly realized that non-blocking sockets were the only sane way to go.

LibCore/AK's API and kernel support for non-blocking sockets was hit or miss but good enough to continue, assuming I could have some simple PRs merged on the way. This is one of them.

Reply:
As mentioned on discord, I don't think that exposing the fd() is the right call for this class, or for the Stream hierarchy in general. If we need to do operations on sockets, we should give them names and add them to the appropriate classes rather than leaking the implementation details to applications.